### ISSUE TITLE
Add Support for Comm objects

### ISSUE THREAD

**Issue by blink1073** [jupyter/jupyter-js-services#24]
I propose the following API for Comm Management.  It uses factory methods returning Promises, in keeping with the Kernel and Session.

http://ipython.org/ipython-doc/dev/development/messaging.html#custom-messages

```typescript
interface INewKernel extends IKernel {
  /**
   * The unhandled message signal for the kernel.
   */
  unhandledMessage: ISignal<IKernel, IKernelMessage>;

  /**
   * An unhandled comm_open message received from the client.
   */
  commOpened: ISignal<IKernel, ICommOpen>;

  /**
   * Send a shell message to the kernel.
   *
   * The future object will yield the result when available.
   */
  sendShellMessage(msg: IKernelMessage, expectReply: boolean): IKernelFuture;

  /**
   * Send a 'comm_info_request', and return the contents of the
   * 'comm_info_reply'.
   */
  commInfo(contents: ICommInfoRequest): Promise<ICommInfoReply>;

  /**
   * Connect to a comm, or create a new one.
   *
   * If a client-side comm already exists, it is returned.
   */
  connectToComm(targetName: string, commId?: string): IComm;
}

/**
 * A client side Comm interface.
 */
export
interface IComm {
  /**
   * The uuid for the comm channel.
   *
   * Read-only
   */
  commId: string;

  /** 
   * The target name for the comm channel.
   *
   * Read-only
   */
  targetName: string;

  /**
   * The onClose handler.
   */
  onClose: (data?: any) => void;

  /**
   * The onMsg handler.
   */
  onMsg: (data: any) => void;

  /**
   * Open a comm with optional data.
   */
  open(data?: any, metadata?: any): IKernelFuture;

  /**
   * Send a comm message to the kernel.
   */
  send(data: any, metadata?: any, buffers?:(ArrayBuffer | ArrayBufferView)[]): IKernelFuture;

  /**
   * Close the comm.
   */
  close(data?: any, metadata?: any): IKernelFuture;
}

/**
 * Contents of `comm_info_reply` message, a mapping of comm ids to target names.
 */
interface ICommInfoReply {
  comms: { [id: string]: string };
}

/**
 * Contents of a `comm_open` message.
 */
export
interface ICommOpen {
  comm_id: string;
  target_name: string;
  data: any;
  target_module?: string;
}
```

Note to self: use this as template for integration test: https://github.com/jupyter/notebook/blob/master/notebook/tests/services/serialize.js

**Comment by blink1073** [comment]
@SylvainCorlay @jdfreder

**Comment by sccolbert** [comment]
lgtm. I assume `INewKernel` is just a placeholder name?

**Comment by blink1073** [comment]
Yeah, I was being cute.

**Comment by blink1073** [comment]
@jasongrout

**Comment by sccolbert** [comment]
@minrk

**Comment by jasongrout** [comment]
@SylvainCorlay, how does this play with your comm info message?

**Comment by blink1073** [comment]
My intent was for `listRunningComms` to use that message under the hood.

**Comment by jasongrout** [comment]
oh, I was thinking listRunningComms would return the existing browser-side comms, not query the server for active comms.  So what exactly do you see listRunningComms doing?  Will it sync to make sure that the active comms are synced between the browser and server?

**Comment by blink1073** [comment]
I was using in place of this [line](https://github.com/ipython/ipywidgets/blob/master/ipywidgets/static/widgets/js/manager.js#L451).  The intent was to get a list of comms running on the server.

**Comment by jdfreder** [comment]
Yay for promises!

If you want to change `comm_id` to `id`, you'll need to change the Python side to `id` too, to keep symmetry.  Alternatively, if you don't want to change the Python, use `commId` in js.

Also, I'm not sure how you intend `onClose` and `onMsg` to work.  Right now, in master, they allow you to register callbacks, but it looks like you've written their signature in the form of the callback, not the function to register the callback.

**Comment by blink1073** [comment]
Ah, I see the intent was to register multiple callbacks, fixed.

**Comment by sccolbert** [comment]
@jdfreder Why are multiple callbacks needed for `onclose` and `onmessage`

**Comment by jdfreder** [comment]
Sorry, I should have been more clear.  When I said:

> they allow you to register callbacks

I meant one callback for each.

> @jdfreder Why are multiple callbacks needed for onclose and onmessage?

Multiple per each aren't needed AFAIK.  ([Python API](https://github.com/ipython/ipykernel/blob/master/ipykernel/comm/comm.py#L132-L148), [current JS API](https://github.com/jupyter/notebook/blob/master/notebook/static/services/kernels/comm.js#L183-L189))

> Ah, I see the intent was to register multiple callbacks, fixed.

Maybe @minrk was anticipating that in the future.  For the most part the JS API mirrors the Python one.

**Comment by blink1073** [comment]
Okay, reverted to the original.

**Comment by sccolbert** [comment]
> Multiple per each aren't needed AFAIK.

Then the original form was correct. The API mirrors websockets. You simply assign your callback handler:

```
comm.onMessage = () => {
  // do stuff
};
```

**Comment by jdfreder** [comment]
You'll want to change the Python API too then, there's no reason for them to not be the same.

**Comment by sccolbert** [comment]
Well, they are different languages. I'd say make the API similar, but don't sacrifice doing it the natural was for Python and the natural way for JS.

**Comment by blink1073** [comment]
I'd say don't change the Python API, since the end result is the same and our intent is to change the front end only as much as possible.

**Comment by jdfreder** [comment]
Part of what makes the current API so attractive is the symmetry.  I agree some of the Pythonisms are an eyesore in Javascript, but I don't think we should dismiss the idea of symmetry so quick.  Especially since this is a network communication API, where devs will be working with both at once.  

I don't see why changing the Python API would cost anymore than changing the JS API, especially small things like `comm_id` -> `id` and removing the helper functions in favor of callback attributes.

**Comment by sccolbert** [comment]
I'll be the last one to dismiss symmetry, but I agree with @blink1073. Let's change one thing at a time, make the JS code as nice to use from JS as possible, then see what makes sense to map back to the server.

**Comment by SylvainCorlay** [comment]
Regarding `ICommManager`'s interface on the creation of comms, there are two cases that must be covered:

- open a new comm -> this creates a comm object and sends a comm open message to the back-end to create the back-end counterpart to it, register the created comm.
- just create a new comm object and register the created comm. This generally follow a comm info request, and we should be able to specify the comm id on creation to match what we got from the back-end.

**Comment by blink1073** [comment]
Yep, those are captured by `startNewComm` and `connectToComm`, respectively.

**Comment by SylvainCorlay** [comment]
ok, so connectToComm also creates a comm object?

**Comment by blink1073** [comment]
yes, it is a factory method for an IComm object

**Comment by blink1073** [comment]
iI just doesn't send the `comm_open` message.

**Comment by SylvainCorlay** [comment]
`startNewComm` could also take an optional comm id, since you can also specify it from the Python side.

**Comment by blink1073** [comment]
Done.

**Comment by blink1073** [comment]
Also added `unregisterTarget` for symmetry.

**Comment by SylvainCorlay** [comment]
Which raises the question : Should there be methods to un-register comms?

**Comment by blink1073** [comment]
My intent was for that to happen when the comm closes.

**Comment by SylvainCorlay** [comment]
I don' t think that we have a usecase for it (or for unregistering targets)

**Comment by minrk** [comment]
I think this looks good as an API. I don't think we need multiple callbacks at this level - if they are needed somewhere down the line, they can be handled via a wrapper, anyway. We don't need to push too hard on the symmetry. Symmetry was part of the design principles, but deviating where it makes sense is fine, as long as there's a reason. @SylvainCorlay, @jasongrout, and @jdfreder will know best what APIs are needed for comm registration in the widget stuff.

**Comment by blink1073** [comment]
ping @SylvainCorlay, I updated the interface to include the `comm_info` messages, do you concur?

**Comment by SylvainCorlay** [comment]
Looks good to me!

**Comment by blink1073** [comment]
sweet, now see if you agree with the implementation :wink:

https://github.com/blink1073/jupyter-js-services/blob/add-comm/src/comm.ts

**Comment by SylvainCorlay** [comment]
Yes, exactly what we want. cf wrapper Promise we in widgets. https://github.com/SylvainCorlay/ipython_widgets/blob/c7ee69352d1401084b09475b13d126198306845b/ipywidgets/static/widgets/js/manager.js#L511

**Comment by jdfreder** [comment]
:+1:

**Comment by blink1073** [comment]
Updated the API to make the `CommManager` completely independent from the `Kernel`, now implemented in #26.

**Comment by blink1073** [comment]
@minrk, @sccolbert, how about the current implementation.  This would be emitted for any message that is not explicitly handled by the `Kernel`.

**Comment by blink1073** [comment]
(Re: `unhandledMessage` signal in `IKernel`)

**Comment by sccolbert** [comment]
Would we want one unhandled message signal per channel?

**Comment by sccolbert** [comment]
+1 on the signal idea though - simplifies implementation

**Comment by blink1073** [comment]
we could just limit it to IOPub and call it `unhandledIOPub`, the channels should all be explicitly handled.

**Comment by sccolbert** [comment]
sounds good to me

**Comment by minrk** [comment]
I like unhandledIOPub

**Comment by blink1073** [comment]
Please see updated API based on discussion this morning.

**Comment by blink1073** [comment]
What is the intended use of target_module here?: https://github.com/jupyter/notebook/blob/master/notebook/static/services/kernels/comm.js#L84
it is not documented anywhere or used anywhere, from what I can tell.

What if we added:
`commOpened: ISignal<IKernel, IKernelMessage>;`, and removed `setCommTargetHandler` entirely?

**Comment by blink1073** [comment]
@SylvainCorlay brought up the point that `target_module` can be used to auto-load Comm targets from nbextensions.  However, it is not documented on https://ipython.org/ipython-doc/dev/development/messaging.html#custom-messages, and is not currently used in the `ipywidget` or `notebook` repos (using a `grep` search).

**Comment by blink1073** [comment]
See updated [description](https://github.com/jupyter/jupyter-js-services/issues/24#issue-105818825).  Comm open messages that do not get loaded via `target_module` would result in a `commOpened` signal, which replaces the target handler registry.

**Comment by blink1073** [comment]
Closed by #26.
